paperscraper/lib.py

```diff
@@ -306,6 +319,96 @@ def parse_semantic_scholar_metadata(paper: dict[str, Any]) -> dict[str, Any]:
+        citation = format_bibtex(bibtex, key, clean=False)
+    if key is None:
+        # get citation by following link
+        # SLOW SLOW Using SerpAPI for this
+        async with session.get(
+            paper["inline_links"]["serpapi_cite_link"]
+            + f"&api_key={os.environ['SERPAPI_API_KEY']}"
+        ) as r:
+            data = await r.json()
+            citation = next(
+                c["snippet"] for c in data["citations"] if c["title"] == "MLA"
+            )
+            bibtex_link = next(
+                c["link"] for c in data["links"] if c["name"] == "BibTeX"
+            )
+        async with session.get(bibtex_link) as r:
+            bibtex = await r.text()
+    return {
+        "citation": citation,
+        "key": key,
+        "bibtex": bibtex,
+        "year": paper["year"],
+        "url": paper["link"],
+        "paperId": paper["paperId"],
+        "doi": paper["externalIds"].get("DOI", None),
+        "citationCount": paper["citationCount"],
+        "title": paper["title"],
+    }
+
+
+async def reconcile_doi(title: str, authors: list[str], session: ClientSession) -> str:
+    # do not want initials
+    authors_query = " ".join([a for a in authors if len(a) > 1])
+    mailto = os.environ.get("CROSSREF_MAILTO", "paperscraper@example.org")
+    # get DOI via crossref
+    url = "https://api.crossref.org/works"
+    params = {
+        "query.title": title,
+        "mailto": mailto,
+        "select": "DOI,score",
+        "rows": "1",
+    }
+    if authors_query:
+        params["query.author"] = authors_query
+    async with session.get(url, params=params) as r:
+        data = await r.json()
+        if data["status"] == "failed":
+            raise DOINotFoundError(f"Could not find DOI for {title}")
+        if (
+            data["message"]["total-results"] == 0
+            or data["message"]["items"][0]["score"] < 0.5  # noqa: PLR2004
+        ):
+            raise DOINotFoundError(f"Could not find DOI for {title}")
+        return data["message"]["items"][0]["DOI"]
+
+
+async def doi_to_bibtex(doi: str, session: ClientSession) -> str:
+    # get DOI via crossref
+    url = f"https://api.crossref.org/works/{doi}/transform/application/x-bibtex"
+    async with session.get(url) as r:
+        data = await r.text()
+    # must make new key
+    key = data.split("{")[1].split(",")[0]
+    new_key = key.replace("_", "")
+    try:
+        author_frag = (
+            data.split("author={")[1]
+            .split("}")[0]
+            .split()[0]
+            .strip()
+            .replace(" and ", "")
+            .replace(",", "")
+        )
+        title_frag = data.split("title={")[1].split("}")[0].split()[0].strip()
+        year_frag = data.split("year={")[1].split("}")[0].split()[0].strip()
+    except IndexError:
+        return data.replace(f"{key}", f"{new_key}")
+    new_key = f"{author_frag}{year_frag}{title_frag}"
+    return data.replace(f"{key}", f"{new_key}")
```

Comment:
I would rather see `str(key)` than `f"{key}"`. Same for `new_key`

Reply:
lol - yea there were more characters previously. looks silly now :)